docs/CONFIGINI.md
## config.ini Options

Look below to find a thorough description of all options and sections available in the `/home/<USER>/.skyscraper/config.ini` configuration file. These options will then be applied whenever running Skyscraper with the `Skyscraper` command.

Please take note that almost all of these options are set at a useful default as documented here (and can therefore be left out) and should only be set if your use case requires it.

!!! warning

    Some users seem to have a habbit of editing the `/home/<USER>/skysource/config.ini` file. _This is not the correct file!_ Please read the first paragraph. :)

Options can be set on four levels; either `[main]`, `[<PLATFORM>]`, `[<FRONTEND>]` or `[<SCRAPER>]`.

Settings in the `[main]` section will always be set regardless of selected platform, frontend and module.

`<PLATFORM>` can be any of the supported platforms (check list with `--help` under the `-p` option), in which case the settings will only be applied while scraping that particular platform.

`<FRONTEND>` can be any of the supported frontends (check list with `--help` under the `-f` option), in which case the settings will only be applied while scraping for that particular frontend.

`<SCRAPER>` can be any of the supported scraping modules (check list with `--help` under the `-s` option), in which case the settings will only be applied while scraping with that particular scraping module.

Each section can have overlapping parameters. In case where a certain option exists in several sections they are prioritized as scraping module first, then frontend, then platform and lastly main. Any commandline option which relates to an configuration setting in `config.ini` has highest precedence, regardless of the other four levels respective sections.

You can find an example config file at `/home/<USER>/.skyscraper/config.ini.example`. This file contains all available options. Just copy the file to `config.ini` and uncomment and edit the ones you wish to use by removing the `#` or `;` in front of the variables. Remember to also uncomment the section the option relates to such as `[main]` or `[amiga]`.

!!! note

    You can set a custom configuration file with the `-c <FILENAME>` command line option. Read more about all available command line options [here](CLIHELP.md).

**Example `config.ini`**:

```ini
[main]
inputFolder="/path/to/your/roms"
cacheFolder="/path/to/your/desired/db/cache/base/folder"
gameListBackup="true"
videos="true"
unattend="true"
verbosity="1"
lang="ja"
region="jp"
cacheMarquees="false"

[amiga]
inputFolder="/specific/path/just/for/amiga/roms"
cacheFolder="/specific/path/just/for/amiga/cache"
minMatch="50"
maxLength="200"

[pegasus]
artworkXml="artwork-pegasus.xml"
maxLength="10000"

[screenscraper]
userCreds="USER:PASSWORD"

```

### Index of Options with Applicable Sections

This is an alphabetical index of all configuration options including the sections they may be applied to.

| Section &rarr;<br>Parameter &darr;                          | `[main]` | `[<PLATFORM>]` | `[<FRONTEND>]` | `[<SCRAPER>]` |
| ----------------------------------------------------------- | :------: | :------------: | :------------: | :-----------: |
| [addExtensions](CONFIGINI.md#addextensions)                 |    Y     |       Y        |                |               |
| [addFolders](CONFIGINI.md#addfolders)                       |          |                |       Y        |               |
| [artworkXml](CONFIGINI.md#artworkxml)                       |    Y     |       Y        |       Y        |               |
| [brackets](CONFIGINI.md#brackets)                           |    Y     |       Y        |       Y        |               |
| [cacheCovers](CONFIGINI.md#cachecovers)                     |    Y     |       Y        |                |       Y       |
| [cacheFolder](CONFIGINI.md#cachefolder)                     |    Y     |       Y        |                |               |
| [cacheMarquees](CONFIGINI.md#cachemarquees)                 |    Y     |       Y        |                |       Y       |
| [cacheRefresh](CONFIGINI.md#cacherefresh)                   |    Y     |                |                |       Y       |
| [cacheResize](CONFIGINI.md#cacheresize)                     |    Y     |       Y        |                |       Y       |
| [cacheScreenshots](CONFIGINI.md#cachescreenshots)           |    Y     |       Y        |                |       Y       |
| [cacheTextures](CONFIGINI.md#cachetextures)                 |    Y     |       Y        |                |       Y       |
| [cacheWheels](CONFIGINI.md#cachewheels)                     |    Y     |       Y        |                |       Y       |
| [cropBlack](CONFIGINI.md#cropblack)                         |    Y     |       Y        |       Y        |               |
| [emulator](CONFIGINI.md#emulator)                           |    Y     |       Y        |       Y        |               |
| [endAt](CONFIGINI.md#endat)                                 |          |       Y        |       Y        |               |
| [excludeFrom](CONFIGINI.md#excludefrom)                     |    Y     |       Y        |                |               |
| [excludePattern](CONFIGINI.md#excludepattern)               |    Y     |       Y        |       Y        |               |
| [extensions](CONFIGINI.md#extensions)                       |          |       Y        |                |               |
| [forceFilename](CONFIGINI.md#forcefilename)                 |    Y     |       Y        |       Y        |               |
| [frontend](CONFIGINI.md#frontend)                           |    Y     |                |                |               |
| [gameListBackup](CONFIGINI.md#gamelistbackup)               |    Y     |                |       Y        |               |
| [gameListFilename](CONFIGINI.md#gamelistfilename)           |          |                |       Y        |               |
| [gameListFolder](CONFIGINI.md#gamelistfolder)               |    Y     |       Y        |       Y        |               |
| [gameListVariants](CONFIGINI.md#gamelistvariants)           |          |                |       Y        |               |
| [hints](CONFIGINI.md#hints)                                 |    Y     |                |                |               |
| [ignoreYearInFilename](CONFIGINI.md#ignoreyearinfilename)   |    Y     |       Y        |                |               |
| [importFolder](CONFIGINI.md#importfolder)                   |    Y     |       Y        |                |               |
| [includeFrom](CONFIGINI.md#includefrom)                     |    Y     |       Y        |                |               |
| [includePattern](CONFIGINI.md#includepattern)               |    Y     |       Y        |       Y        |               |
| [innerBracketsReplace](CONFIGINI.md#innerbracketsreplace)   |    Y     |                |                |               |
| [innerParenthesesReplace](CONFIGINI.md#innerparenthesesreplace) |  Y   |                |                |               |
| [inputFolder](CONFIGINI.md#inputfolder)                     |    Y     |       Y        |                |               |
| [interactive](CONFIGINI.md#interactive)                     |    Y     |       Y        |                |       Y       |
| [jpgQuality](CONFIGINI.md#jpgquality)                       |    Y     |       Y        |                |       Y       |
| [keepDiscInfo](CONFIGINI.md#keepdiscinfo)                   |    Y     |       Y        |                |               |
| [lang](CONFIGINI.md#lang)                                   |    Y     |       Y        |                |               |
| [langPrios](CONFIGINI.md#langprios)                         |    Y     |       Y        |                |               |
| [launch](CONFIGINI.md#launch)                               |    Y     |       Y        |       Y        |               |
| [manuals](CONFIGINI.md#manuals)                             |    Y     |       Y        |                |               |
| [maxFails](CONFIGINI.md#maxfails)                           |    Y     |                |                |               |
| [maxLength](CONFIGINI.md#maxlength)                         |    Y     |       Y        |       Y        |       Y       |
| [mediaFolder](CONFIGINI.md#mediafolder)                     |    Y     |       Y        |       Y        |               |
| [mediaFolderHidden](CONFIGINI.md#mediafolderhidden)         |          |                |       Y        |               |
| [minMatch](CONFIGINI.md#minmatch)                           |    Y     |       Y        |                |       Y       |
| [nameTemplate](CONFIGINI.md#nametemplate)                   |    Y     |       Y        |                |               |
| [onlyMissing](CONFIGINI.md#onlymissing)                     |    Y     |       Y        |                |       Y       |
| [platform](CONFIGINI.md#platform)                           |    Y     |                |                |               |
| [pretend](CONFIGINI.md#pretend)                             |    Y     |       Y        |                |               |
| [region](CONFIGINI.md#region)                               |    Y     |       Y        |                |               |
| [regionPrios](CONFIGINI.md#regionprios)                     |    Y     |       Y        |                |               |
| [relativePaths](CONFIGINI.md#relativepaths)                 |    Y     |       Y        |                |               |
| [scummIni](CONFIGINI.md#scummini)                           |    Y     |                |                |               |
| [skipped](CONFIGINI.md#skipped)                             |    Y     |       Y        |       Y        |               |
| [spaceCheck](CONFIGINI.md#spacecheck)                       |    Y     |                |                |               |
| [startAt](CONFIGINI.md#startat)                             |          |       Y        |       Y        |               |
| [subdirs](CONFIGINI.md#subdirs)                             |    Y     |       Y        |                |               |
| [symlink](CONFIGINI.md#symlink)                             |    Y     |       Y        |       Y        |               |
| [theInFront](CONFIGINI.md#theinfront)                       |    Y     |       Y        |       Y        |               |
| [threads](CONFIGINI.md#threads)                             |    Y     |       Y        |                |       Y       |
| [tidyDesc](CONFIGINI.md#tidydesc)                           |    Y     |       Y        |                |       Y       |
| [unattend](CONFIGINI.md#unattend)                           |    Y     |       Y        |       Y        |       Y       |
| [unattendSkip](CONFIGINI.md#unattendskip)                   |    Y     |       Y        |       Y        |       Y       |
| [unpack](CONFIGINI.md#unpack)                               |    Y     |       Y        |                |               |
| [userCreds](CONFIGINI.md#usercreds)                         |          |                |                |       Y       |
| [verbosity](CONFIGINI.md#verbosity)                         |    Y     |       Y        |       Y        |               |
| [videoConvertCommand](CONFIGINI.md#videoconvertcommand)     |    Y     |                |                |       Y       |
| [videoConvertExtension](CONFIGINI.md#videoconvertextension) |    Y     |                |                |       Y       |
| [videoPreferNormalized](CONFIGINI.md#videoprefernormalized) |          |                |                |       Y       |
| [videos](CONFIGINI.md#videos)                               |    Y     |       Y        |       Y        |       Y       |
| [videoSizeLimit](CONFIGINI.md#videosizelimit)               |    Y     |       Y        |                |       Y       |

#### inputFolder

Sets the rom input folder. By default Skyscraper will look for roms in the `/home/<USER>/RetroPie/roms/<PLATFORM>` folder. If your roms are located in a non-default location, you can set the input path using this option.

!!! note

    If this is set in the `[main]` section it will automatically add `/<PLATFORM>` to the end of the path. If you want better control consider adding it to a `[<PLATFORM>]` section instead where it will be used as is.

Default value: `/home/<USER>/RetroPie/roms/<PLATFORM>`  
Allowed in sections: `[main]`, `[<PLATFORM>]`

---

#### gameListFolder

Sets the game list export folder. By default Skyscraper exports the game list to the same directory as the rom input folder. This enables you to change that to a non-default location.

!!! note

    If this is set in the `[main]` or `[<FRONTEND>]` section it will automatically add `/<PLATFORM>` to the end of the path. If you want better control consider adding it to a `[<PLATFORM>]` section instead where it will be used as is.

Default value: `/home/<USER>/RetroPie/roms/<PLATFORM>`  
Allowed in sections: `[main]`, `[<PLATFORM>]`, `[<FRONTEND>]`

---

#### gameListFilename

Override the game list filename. This enables you to set a filename different from the default provided by the frontend modules. For example you can specify a custom filename for the gamelist, which is particularly useful if you are using EmuDeck with the Pegasus frontend: This enhancement helps prevent duplicate entries in Pegasus by allowing users to generate a `metadata.txt` file instead of the default `metadata.pegasus.txt`.

Default value: depends on [frontend](FRONTENDS.md)  
Allowed in sections: `[<FRONTEND>]`

---

#### gameListBackup

When set to true Skyscraper will create a backup of the existing game list each time it is run in game list generation mode (this mode is enabled by leaving out the `-s MODULE` option). The filename and path of the backup will be equal to the existing game list but have a timestamp appended to the name of the format `-yyyyMMdd-hhmmss` (eg. `gamelist.xml-20200530-115900`).

Default value: `false`  
Allowed in sections: `[main]`, `[<FRONTEND>]`

---


docs/CHANGELOG.md
## Changes

This page summarizes the changes of each Skyscraper release, a [changlog for
humans](https://keepachangelog.com).

### Version 3.14.0 (2024-12-08)

- Added: Support for [XDG Base Directories](XDG.md), thanks for the suggestion
  @ASHGOLDOFFICIAL. 
- Added: Option to allow any delimiter between consecutive brackets and
  parentheses in gamelist title. See
  [`innerBracketsReplace`](CONFIGINI.md#innerbracketsreplace) for examples.
  Thanks for the suggestion, @retrobit.
- Added: Option to retain disc numbering from game filename, when no other
  bracket information is requested. See
  [`keepDiscInfo`](CONFIGINI.md#keepdiscinfo) for details. Thanks, @maxexcloo!
- Added: Option to override year comparison during scraping, if year is present
  in game filename. See
  [`ignoreYearInFilename`](CONFIGINI.md#ignoreyearinfilename).
- Added: [Platform 'Fujitsu
  FM-Towns'](https://github.com/Gemba/skyscraper/pull/95/files). Manually update
  your `peas.json` and `platformid_map.csv` to make use of it.
- Added: Option `--hint`, it shows a random Tip of the Day.
- Updated: Skyscraper's hardcoded `/home/<USER>` replaced with the actual user's
  home directory in messages. Thanks for highlighting it on the Mac,
  @cdaters 
- Changed: When an invalid scrape module is provided with `-s` Skyscraper now
  exits. Before this change Skyscraper failed back to cache scraping silently.
- Updated: Check on RetroPie if an existing Skyscraper installation is updated
  at least with RetroPie-Setup 4.8.6 onwards to have the configurable platform
  information deployed (`peas.json`) and provide a remediation to the user, if
  this is violated
- Updated: A downloaded `whdload.xml` file for platform Amiga will be not
  downloaded again until the server indicates. However, manually removing
  `/home/<USER>/.skyscraper/whdload_cached_etag.txt` will force a new download.
- Fixed: Performing Ctrl-C in `--cache edit` mode will now dismiss any changes
  made instead of persisting them
- Fixed: Game rating calculation in Openretro scraping module

### Version 3.13.0 (2024-11-06)

- Added: Option to provide user file `peas_local.json` (same format as
  `peas.json`), to extend platform information or overwrite existing platform
  information
- Updated: `*_platform.json` files as reference for supported platforms of
  various scrapers
- Added: Support for Vircon32 platform in `peas.json`. Thanks for hinting,
  @vircon32
- Removed: `scrapers` entries in `peas.json`, as it did not provide any use.
- Removed: Deprecated flags and options: `includeFiles` superseded by
  `includePattern`; `excludeFiles` superseded by `excludePattern`;
  `gamelistFolder` superseded by `gameListFolder`; `fromfile` superseded by
  `includefrom`. These were deprecated since v3.7.0.

### Version 3.12.0 (2024-07-01)

- Added: Support for scraping of PDF manuals (for scrape modules screenscraper,
  import and esgamelist) and gamelist output with these manuals for frontends
  (ES-DE Frontend and some EmulationStation variants). See configurations
  options [`manuals=true`](CONFIGINI.md#manuals) and
  [`gameListVariants=enable-manuals`](CONFIGINI.md#gamelistvariants). Thanks for
  the initial PR, @pandino
- Added: For frontend ES-DE, evaluate environment variable `ESDE_APPDATA_DIR` if
  present. Thanks for the hint, @ASHGOLDOFFICIAL
- Added: Use also release year as hint on user interaction. This is useful when
  in interactive mode and more than one game with the same name is found.
  Skyscraper can be guided to prefer a specific game when the release year is
  added in parenthesis as part of the ROM name (or alias in `aliasMap.csv`).
  Verbose info in [#59](https://github.com/Gemba/skyscraper/pull/59). Thanks,
  @mjkaye
- Changed: Persistent config option `onlymissing` for counting and scraping only
  games which do not have any game data in the cache. This is a commodity config
  option to the already existing flag with the same name. Plus: If you use a
  scraper with a scraping limit for games to be scraped at once (e.g. MobyGames)
  you may stay below that limit. Documentation [here](CONFIGINI.md#onlymissing).
  Thanks for the suggestion, @sleve_mcdichael
- Update: Valid extensions (= `formats` in Skyscraper's `peas.json` file) with
  info from RetroPie's `platform.cfg` (commit
  [`5e0ab1f`](https://github.com/RetroPie/RetroPie-Setup/blob/5e0ab1f85994cbb51eb5539d2a7592a3578c15b8/platforms.cfg))
- Fixed: The [update
  script](https://github.com/Gemba/skyscraper/blob/master/update_skyscraper.sh)
  for recent MacOS versions. Thanks, @calumbrodie 

### Version 3.11.0 (2024-04-15)

- Added: Support for EmulationStation Desktop Edition (ES-DE Frontend). Use
  [`frontend=esde`](CONFIGINI.md#frontend) in `config.ini` and see
  [documentation](FRONTENDS.md#emulationstation-desktop-edition-es-de) on the
  default settings. Thanks for the hints and for testing, @maxexcloo, @Nargash
- Added: Entries in
  [`aliasMap.csv`](https://github.com/Gemba/skyscraper/blob/master/aliasMap.csv)
  are now also applicable for Screenscraper. Thanks, @retrobit.
- Added: Enhanced game name detection for ScummVM platform.
- Changed: In module Screenscraper, in some cases if multiple media types are
  queried (NB: most queries are single-typed), now the first type is selected
  with the first matching region. Only if this does exists in any provided
  region, try the next type with all provided regions (in short: type has
  precedence over region). Previously the regions have been tried to match on
  any media type (=region had precedence over type), which resulted in picking
  the alternative type. This approach gave less suitable media files, especially
  for screenshots (type:`ss`) and screenshot from title screen (type:
  `sstitle`).
- Fixed: Some corner-case bugs fixed, thanks to all reporters!

### Version 3.10.0 (2024-02-10)

- Feature: Preserve existing `<folder/>` nodes in gamelist or create skeleton

src/settings.h
/***************************************************************************
 *            settings.h
 *
 *  Wed Jun 14 12:00:00 CEST 2017
 *  Copyright 2017 Lars Muldjord
 *  muldjordlars@gmail.com
 ****************************************************************************/
/*
 *  This file is part of skyscraper.
 *
 *  skyscraper is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  skyscraper is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with skyscraper; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
 */

#ifndef SETTINGS_H
#define SETTINGS_H

#include "config.h"

#include <QCommandLineParser>
#include <QImage>
#include <QMap>
#include <QObject>
#include <QPair>
#include <QSettings>

struct Settings {
    bool isMatchOneScraper() {
        const QStringList matchOneScraper = QStringList(
            {"cache", "screenscraper", "arcadedb", "esgamelist", "import"});
        return matchOneScraper.contains(scraper);
    }
    QString currentDir = "";

    QString cacheFolder = "";
    QString gameListFileString = "";
    QString skippedFileString = "";
    QString configFile = "";
    QString frontend = "";
    // For use with 'emulator' with Attractmode and 'command' for Pegasus
    QString frontendExtra = "";
    QString platform = "";
    bool arcadePlatform;
    QString scraper = "";
    QString userCreds = "";
    QString igdbToken = "";
    QString inputFolder = "";
    QString gameListFolder = "";
    QString gameListFilename = "";
    QString mediaFolder = "";
    // Next two only relevant for EmulationStation/ES-DE
    bool mediaFolderHidden = false; // EmulationStation only
    bool addFolders = false;        // EmulationStation and ES-DE
    QString screenshotsFolder = "";
    QString coversFolder = "";
    QString wheelsFolder = "";
    QString marqueesFolder = "";
    QString texturesFolder = "";
    QString videosFolder = "";
    QString manualsFolder = "";
    QString importFolder = Config::getSkyFolder(Config::SkyFolderType::IMPORT);
    QString nameTemplate = "";
    int doneThreads = 0;
    int threads = 4;
    bool threadsSet = false;
    int minMatch = 65;
    bool minMatchSet = false;
    int maxLength = 2500;
    bool brackets = true;
    bool keepDiscInfo = false;
    bool refresh = false;
    QString cacheOptions = "";
    bool cacheResize = true;
    int jpgQuality = 95;
    bool subdirs = true;
    bool onlyMissing = false;
    QString startAt = "";
    QString endAt = "";
    bool pretend = false;
    bool unattend = false;
    bool unattendSkip = false;
    bool interactive = false;
    bool forceFilename = false;
    bool stats = false;
    bool hints = true;
    int verbosity = 0;
    int maxFails = 42;
    bool skipped = false;
    bool tidyDesc = true;
    bool ignoreYearInFilename = false;
    QString artworkConfig = "";
    QByteArray artworkXml = "";
    QString excludePattern = "";
    QString includePattern = "";
    QString includeFrom = "";
    QString excludeFrom = "";
    bool relativePaths = false;
    QString extensions = "";
    QString addExtensions = "";
    bool unpack = false;
    bool theInFront = false;
    bool gameListBackup = false;
    bool preserveOldGameList = true;
    bool spaceCheck = true;
    QString scummIni = "";

    int romLimit = -1;

    bool manuals = false;
    QString gameListVariants = "";
    bool videos = false;
    bool videoPreferNormalized = true;
    int videoSizeLimit = 100 * 1000 * 1000;
    QString videoConvertCommand = "";
    QString videoConvertExtension = "";
    bool symlink = false;
    bool skipExistingVideos = false;
    bool cacheCovers = true;
    bool skipExistingCovers = false;
    bool cacheScreenshots = true;
    bool skipExistingScreenshots = false;
    // If true, will crop away black borders around screenshots
    bool cropBlack = true;
    bool cacheWheels = true;
    bool skipExistingWheels = false;
    bool cacheMarquees = true;
    bool skipExistingMarquees = false;
    bool skipExistingTextures = false;
    bool cacheTextures = true;
    bool skipExistingManuals = false;
    QString innerBracketsReplace = "";
    QString innerParenthesesReplace = "";

    QString user = "";
    QString password = "";
    QString lang = "";
    QString region = "";
    QString langPriosStr = "";
    QString regionPriosStr = "";

    QString searchName = "";

    QMap<QString, QImage> resources;
    QMap<QString, QString> mameMap;
    QMap<QString, QString> aliasMap;
    QMap<QString, QPair<QString, QString>> whdLoadMap;

    QList<QString> regionPrios;
    QList<QString> langPrios;
};

class RuntimeCfg : public QObject {
    Q_OBJECT

public:
    enum CfgType { MAIN = 1, PLATFORM = 2, FRONTEND = 4, SCRAPER = 8 };

    RuntimeCfg(Settings *config, const QCommandLineParser *parser);
    ~RuntimeCfg();

    void applyConfigIni(CfgType type, QSettings *settings, bool &inputFolderSet,
                        bool &gameListFolderSet, bool &mediaFolderSet);

    void applyCli(bool &inputFolderSet, bool &gameListFolderSet,
                  bool &mediaFolderSet);

    QString concatPath(QString absPath, QString platformFolder);
    bool validateFrontend(const QString &providedFrontend);

private:
    void setFlag(const QString flag);
    QSet<QString> getKeys(CfgType type);
    QStringList parseFlags();
    void reportInvalidPlatform();

    Settings *config;
    const QCommandLineParser *parser;

    QMap<QString, QPair<QString, int>> params = {
        // clang-format off
        {"addExtensions",           QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"addFolders",              QPair<QString, int>("bool",                                     CfgType::FRONTEND                    )},
        {"artworkXml",              QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"brackets",                QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"cacheCovers",             QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )},
        {"cacheFolder",             QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"cacheMarquees",           QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )},
        {"cacheRefresh",            QPair<QString, int>("bool", CfgType::MAIN |                                         CfgType::SCRAPER )},
        {"cacheResize",             QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )},
        {"cacheScreenshots",        QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )},
        {"cacheTextures",           QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )},
        {"cacheWheels",             QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )},
        {"cropBlack",               QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"emulator",                QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"endAt",                   QPair<QString, int>("str",                  CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"excludeFrom",             QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"excludePattern",          QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"extensions",              QPair<QString, int>("str",                  CfgType::PLATFORM                                        )},
        {"forceFilename",           QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"frontend",                QPair<QString, int>("str",  CfgType::MAIN                                                            )},
        {"gameListBackup",          QPair<QString, int>("bool", CfgType::MAIN |                     CfgType::FRONTEND                    )},
        {"gameListFilename",        QPair<QString, int>("str",                                      CfgType::FRONTEND                    )},
        {"gameListFolder",          QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"gameListVariants",        QPair<QString, int>("str",                                      CfgType::FRONTEND                    )},
        {"hints",                   QPair<QString, int>("bool", CfgType::MAIN                                                            )},
        {"ignoreYearInFilename",    QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"importFolder",            QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"includeFrom",             QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"includePattern",          QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"innerBracketsReplace",    QPair<QString, int>("str",  CfgType::MAIN                                                            )},
        {"innerParenthesesReplace", QPair<QString, int>("str",  CfgType::MAIN                                                            )},
        {"inputFolder",             QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"interactive",             QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )},
        {"jpgQuality",              QPair<QString, int>("int",  CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )},        
        {"keepDiscInfo",            QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"lang",                    QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"langPrios",               QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"launch",                  QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"manuals",                 QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"maxFails",                QPair<QString, int>("int",  CfgType::MAIN                                                            )},
        {"maxLength",               QPair<QString, int>("int",  CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND | CfgType::SCRAPER )},
        {"mediaFolder",             QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"mediaFolderHidden",       QPair<QString, int>("bool",                                     CfgType::FRONTEND                    )},
        {"minMatch",                QPair<QString, int>("int",  CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )},
        {"nameTemplate",            QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"onlyMissing",             QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )},
        {"platform",                QPair<QString, int>("str",  CfgType::MAIN                                                            )},
        {"pretend",                 QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"region",                  QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"regionPrios",             QPair<QString, int>("str",  CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"relativePaths",           QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"scummIni",                QPair<QString, int>("str",  CfgType::MAIN                                                            )},
        {"skipped",                 QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"spaceCheck",              QPair<QString, int>("bool", CfgType::MAIN                                                            )},
        {"startAt",                 QPair<QString, int>("str",                  CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"subdirs",                 QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"symlink",                 QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"theInFront",              QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"threads",                 QPair<QString, int>("int",  CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )},
        {"tidyDesc",                QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )},
        {"unattend",                QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND | CfgType::SCRAPER )},
        {"unattendSkip",            QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND | CfgType::SCRAPER )},
        {"unpack",                  QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM                                        )},
        {"userCreds",               QPair<QString, int>("str",                                                          CfgType::SCRAPER )},
        {"verbosity",               QPair<QString, int>("int",  CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND                    )},
        {"videoConvertCommand",     QPair<QString, int>("str",  CfgType::MAIN |                                         CfgType::SCRAPER )},
        {"videoConvertExtension",   QPair<QString, int>("str",  CfgType::MAIN |                                         CfgType::SCRAPER )},
        {"videoPreferNormalized",   QPair<QString, int>("bool",                                                         CfgType::SCRAPER )},
        {"videos",                  QPair<QString, int>("bool", CfgType::MAIN | CfgType::PLATFORM | CfgType::FRONTEND | CfgType::SCRAPER )},
        {"videoSizeLimit",          QPair<QString, int>("int",  CfgType::MAIN | CfgType::PLATFORM |                     CfgType::SCRAPER )}
        // clang-format on
    };
};

#endif // SETTINGS_H
